data/summary/types/cgm.go

```diff
@@ -0,0 +1,272 @@
+
+	// actual values
+	TimeCGMUsePercent *float64 `json:"timeCGMUsePercent" bson:"timeCGMUsePercent"`
+	TimeCGMUseMinutes int      `json:"timeCGMUseMinutes" bson:"timeCGMUseMinutes"`
+	TimeCGMUseRecords int      `json:"timeCGMUseRecords" bson:"timeCGMUseRecords"`
+
+	AverageGlucose             *Glucose `json:"averageGlucose" bson:"avgGlucose"`
+	GlucoseManagementIndicator *float64 `json:"glucoseManagementIndicator" bson:"glucoseManagementIndicator"`
+
+	TimeInTargetPercent *float64 `json:"timeInTargetPercent" bson:"timeInTargetPercent"`
+	TimeInTargetMinutes int      `json:"timeInTargetMinutes" bson:"timeInTargetMinutes"`
+	TimeInTargetRecords int      `json:"timeInTargetRecords" bson:"timeInTargetRecords"`
+
+	TimeInLowPercent *float64 `json:"timeInLowPercent" bson:"timeInLowPercent"`
+	TimeInLowMinutes int      `json:"timeInLowMinutes" bson:"timeInLowMinutes"`
+	TimeInLowRecords int      `json:"timeInLowRecords" bson:"timeInLowRecords"`
+
+	TimeInVeryLowPercent *float64 `json:"timeInVeryLowPercent" bson:"timeInVeryLowPercent"`
+	TimeInVeryLowMinutes int      `json:"timeInVeryLowMinutes" bson:"timeInVeryLowMinutes"`
+	TimeInVeryLowRecords int      `json:"timeInVeryLowRecords" bson:"timeInVeryLowRecords"`
+
+	TimeInHighPercent *float64 `json:"timeInHighPercent" bson:"timeInHighPercent"`
+	TimeInHighMinutes int      `json:"timeInHighMinutes" bson:"timeInHighMinutes"`
+	TimeInHighRecords int      `json:"timeInHighRecords" bson:"timeInHighRecords"`
+
+	TimeInVeryHighPercent *float64 `json:"timeInVeryHighPercent" bson:"timeInVeryHighPercent"`
+	TimeInVeryHighMinutes int      `json:"timeInVeryHighMinutes" bson:"timeInVeryHighMinutes"`
+	TimeInVeryHighRecords int      `json:"timeInVeryHighRecords" bson:"timeInVeryHighRecords"`
+}
+
+type CGMPeriods map[string]CGMPeriod
+
+type CGMStats struct {
+	Periods    CGMPeriods                             `json:"periods" bson:"periods"`
+	Buckets    Buckets[CGMBucketData, *CGMBucketData] `json:"buckets" bson:"buckets"`
+	TotalHours int                                    `json:"totalHours" bson:"totalHours"`
+}
+
+func (*CGMStats) GetType() string {
+	return SummaryTypeCGM
+}
+
+func (s *CGMStats) Init() {
+	s.Buckets = make(Buckets[CGMBucketData, *CGMBucketData], 0)
+	s.Periods = make(map[string]CGMPeriod)
+	s.TotalHours = 0
+}
+
+func (s *CGMStats) GetBucketsLen() int {
+	return len(s.Buckets)
+}
+
+func (s *CGMStats) GetBucketDate(i int) time.Time {
+	return s.Buckets[i].Date
+}
+
+func (s *CGMStats) Update(userData any) error {
+	userDataTyped := userData.([]*glucoseDatum.Glucose)
+	err := AddData(s.Buckets, userDataTyped)
+	if err != nil {
+		return err
+	}
+
+	s.CalculateSummary()
+
+	return nil
+}
+
+func (B *CGMBucketData) CalculateStats(r interface{}, lastRecordTime *time.Time) error {
+	dataRecord := r.(*glucoseDatum.Glucose)
+	var normalizedValue float64
+	var duration int
+
+	// duration has never been calculated, use current record's duration for this cycle
+	if duration == 0 {
```

Comment:
this originally was meant to use the previous record, and this weird method is just a shim to make it work.
It should be revisited, let me come up with a better solution, as this can result in some math errors with competing sources of very-different duration.

Reply:
made it work as intended again.